### PR TITLE
FlatAHB Dump: Don't use new GUID for otherwise unchanged lines

### PR DESCRIPTION
**Pull request by hf-kklein**
This should reduce the technical noise e.g. in https://github.com/Hochfrequenz/machine-readable_anwendungshandbuecher/commit/ae8cfe6001193e79fe80295dd60c4a385c873845

### CODE REVIEW

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py [hunk 1]
```diff
@@ -286,12 +287,31 @@ def dump_flatahb_json(self, output_directory_path: Path) -> None:
...
+            for line, existing_line in zip(flat_ahb.lines, existing_flat_ahb.lines):
+                line_copy = ahb_line_schema.load(ahb_line_schema.dump(line))
+                existing_line_copy = ahb_line_schema.load(ahb_line_schema.dump(existing_line))
+                line_copy.guid = None
+                existing_line_copy.guid = None
+                if line_copy == existing_line_copy:
+                    # keep the same guid if lines didn't change
+                    line.guid = existing_line.guid
+                else:
+                    dump_equals_existing_file_except_for_guids = False
```
Comment (hf-krechan) [inline src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:310]: when do we need this part?

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py [hunk 2]
```diff
@@ -286,12 +287,31 @@ def dump_flatahb_json(self, output_directory_path: Path) -> None:
...
+        if dump_equals_existing_file_except_for_guids:
```
Comment (hf-kklein) [inline src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:311]: an dieser stelle

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py (hunk 1 above)
Comment (hf-kklein) [reply src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:310]: https://github.com/Hochfrequenz/kohlrahbi/pull/158/files#r1271812201

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py [hunk 3]
```diff
@@ -286,12 +287,31 @@ def dump_flatahb_json(self, output_directory_path: Path) -> None:
...
-        with open(
-            flatahb_output_directory_path / f"{self.meta_data.pruefidentifikator}.json", "w", encoding="utf-8"
-        ) as file:
+        flat_ahb = self.convert_to_flat_ahb()
+
+        file_path = flatahb_output_directory_path / f"{self.meta_data.pruefidentifikator}.json"
+        if file_path.exists():
+            with open(file_path, "r", encoding="utf-8") as file:
+                existing_flat_ahb = FlatAnwendungshandbuchSchema().load(json.load(file))
+        dump_equals_existing_file_except_for_guids = flat_ahb.meta == existing_flat_ahb.meta and len(
```
Comment (hf-krechan) [inline src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:296]: `dump_equals_existing_file_except_for_guids` I think we can improve this name.
It is not so clear at the first reading.

Can you explain what this variable controls?

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py (hunk 2 above)
Comment (hf-krechan) [inline src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:311]: This is the same condition like in line 299. Is this by design?

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py [hunk 4]
```diff
@@ -286,12 +287,31 @@ def dump_flatahb_json(self, output_directory_path: Path) -> None:
...
+            flat_ahb.lines
+        ) == len(existing_flat_ahb.lines)
+        if dump_equals_existing_file_except_for_guids:
+            ahb_line_schema = AhbLineSchema()
+            for line, existing_line in zip(flat_ahb.lines, existing_flat_ahb.lines):
+                line_copy = ahb_line_schema.load(ahb_line_schema.dump(line))
```
Comment (hf-krechan) [inline src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:329]: This is a C# pattern for deep copies isn't it?
I thought in Python there are methods to create deep copies.
* asking chatGPT *
ah yes there is a standard lib called `copy`
See https://chat.openai.com/share/4ad4c1ad-6f13-4a21-a0e3-aaf75e17c556

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py [hunk 5]
```diff
@@ -27,6 +27,38 @@
...
 
+def _lines_are_equal_when_ignoring_guid(line1: AhbLine, line2: AhbLine) -> bool:
+    """
+    returns true iff the line1 and line2 are equal except for their guid
+    """
+    line1_copy = copy.deepcopy(line1)
+    line2_copy = copy.deepcopy(line2)
+    line1_copy.guid = None
+    line2_copy.guid = None
+    return line1_copy == line2_copy
```
Comment (hf-krechan) [inline src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:38]: uh nice. great that the AhbLine class supports is_equal out of the box :)

**Review by hf-krechan — APPROVED** [review]
Sweet I like it :)

File: src/kohlrahbi/unfoldedahb/unfoldedahbtable.py (hunk 5 above)
Comment (hf-kklein) [reply src/kohlrahbi/unfoldedahb/unfoldedahbtable.py:38]: well, that's one purpose of attrs, pydantic, dataclasses and alike

**Pull request merged**